File: backend/api/routers/automation.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query

from backend.api.dependencies import get_container
from backend.api.schemas.automation import (
    ApprovalDecisionRequest,
    CreateAutomationWorkflowRequest,
    ErrorPayload,
    InstantiateRecipeRequest,
    ResponseEnvelope,
    RunAutomationRequest,
    SetAutomationEnabledRequest,
    SetGlobalAutomationRequest,
)
from backend.application.automation import AutomationApplicationError
from backend.domain.shared_kernel import ProjectId
from backend.infrastructure.di import ApplicationContainer


router = APIRouter(prefix="/api/v1", tags=["automation"])


@router.get("/automation/settings", response_model=ResponseEnvelope)
def get_global_automation_settings(container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    return _success(container.create_automation_service().get_global_settings())


@router.patch("/automation/settings", response_model=ResponseEnvelope)
def set_global_automation_settings(
    request: SetGlobalAutomationRequest,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    return _success(container.create_automation_service().set_global_enabled(enabled=request.global_enabled))


@router.get("/projects/{project_id}/automation/workflows", response_model=ResponseEnvelope)
def list_automation_workflows(project_id: str, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        return _success(container.create_automation_service().list_workflows(ProjectId(project_id)))
    except AutomationApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/automation/workflows", response_model=ResponseEnvelope)
def create_automation_workflow(
    project_id: str,
    request: CreateAutomationWorkflowRequest,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(
            container.create_automation_service().create_workflow(
                ProjectId(project_id),
                name=request.name,
                description=request.description,
                trigger_type=request.trigger_type,
                trigger_config=request.trigger_config,
                conditions=[item.model_dump() for item in request.conditions] if request.conditions else None,
                actions=[item.model_dump() for item in request.actions],
                schedule_interval_seconds=request.schedule_interval_seconds,
                is_enabled=request.is_enabled,
            )
        )
    except AutomationApplicationError as error:
        return _failure(error)


@router.patch("/projects/{project_id}/automation/workflows/{workflow_id}", response_model=ResponseEnvelope)
def set_automation_workflow_enabled(
    project_id: str,
    workflow_id: str,
    request: SetAutomationEnabledRequest,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(
            container.create_automation_service().set_workflow_enabled(
                ProjectId(project_id), workflow_id, is_enabled=request.is_enabled
            )
        )
    except AutomationApplicationError as error:
        return _failure(error)


@router.get("/projects/{project_id}/automation/runs", response_model=ResponseEnvelope)
def list_automation_runs(
    project_id: str,
    limit: int = Query(default=50, ge=1, le=200),
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(container.create_automation_service().list_runs(ProjectId(project_id), limit=limit))
    except AutomationApplicationError as error:
        return _failure(error)


@router.get("/projects/{project_id}/automation/runs/{run_id}", response_model=ResponseEnvelope)
def get_automation_run(
    project_id: str,
    run_id: str,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(container.create_automation_service().get_run(ProjectId(project_id), run_id))
    except AutomationApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/automation/workflows/{workflow_id}/run", response_model=ResponseEnvelope)
def run_automation_now(
    project_id: str,
    workflow_id: str,
    request: RunAutomationRequest,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(
            container.create_automation_service().run_now(
                ProjectId(project_id), workflow_id, idempotency_key=request.idempotency_key
            )
        )
    except AutomationApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/automation/run-steps/{step_id}/undo", response_model=ResponseEnvelope)
def undo_automation_run_step(
    project_id: str,
    step_id: str,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(container.create_automation_service().undo_run_step(ProjectId(project_id), step_id))
    except AutomationApplicationError as error:
        return _failure(error)


@router.get("/automation/recipes", response_model=ResponseEnvelope)
def list_automation_recipes(container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    return _success(container.create_automation_service().list_recipes())


@router.post("/projects/{project_id}/automation/recipes/{recipe_key}", response_model=ResponseEnvelope)
def instantiate_automation_recipe(
    project_id: str,
    recipe_key: str,
    request: InstantiateRecipeRequest,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(
            container.create_automation_service().instantiate_recipe(
                ProjectId(project_id),
                recipe_key,
                params=request.params,
                is_enabled=request.is_enabled,
            )
        )
    except AutomationApplicationError as error:
        return _failure(error)


@router.get("/projects/{project_id}/automation/recipe-instances", response_model=ResponseEnvelope)
def list_recipe_instances(project_id: str, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        return _success(container.create_automation_service().list_recipe_instances(ProjectId(project_id)))
    except AutomationApplicationError as error:
        return _failure(error)


@router.get("/projects/{project_id}/automation/approvals/pending", response_model=ResponseEnvelope)
def list_pending_approvals(project_id: str, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        return _success(container.create_automation_service().list_pending_approvals(ProjectId(project_id)))
    except AutomationApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/automation/runs/{run_id}/approvals/{approval_id}/approve", response_model=ResponseEnvelope)
def approve_automation_run(
    project_id: str,
    run_id: str,
    approval_id: str,
    request: ApprovalDecisionRequest,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(
            container.create_automation_service().approve_run(
                ProjectId(project_id), run_id, approval_id, decided_by=request.decided_by
            )
        )
    except AutomationApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/automation/runs/{run_id}/approvals/{approval_id}/reject", response_model=ResponseEnvelope)
def reject_automation_run(
    project_id: str,
    run_id: str,
    approval_id: str,
    request: ApprovalDecisionRequest,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(
            container.create_automation_service().reject_run(
                ProjectId(project_id), run_id, approval_id, reason=request.reason, decided_by=request.decided_by
            )
        )
    except AutomationApplicationError as error:
        return _failure(error)


def _success(data: dict | list[dict]) -> ResponseEnvelope:
    return ResponseEnvelope(ok=True, data=data)


def _failure(error: AutomationApplicationError) -> ResponseEnvelope:
    return ResponseEnvelope(ok=False, error=ErrorPayload(code=error.code.value, message=str(error)))

File: frontend/src/api/automation.ts
import { jsonRequest, requestBackend, type BackendResponse } from "./backend";

export interface GlobalAutomationSettings {
  global_enabled: boolean;
}

export interface AutomationRecipe {
  recipe_key: string;
  name: string;
  description: string;
  trigger_type: string;
  required_capabilities: string[];
  deferred_capabilities: string[];
  instantiation_status: string;
  actions: Array<{
    action_type: string;
    execution_tier: string;
    safety_class: string;
    capability_id: string | null;
    deferred: boolean;
  }>;
}

export interface AutomationWorkflow {
  automation_workflow_id: string;
  project_id: string;
  name: string;
  description: string | null;
  is_enabled: boolean;
  current_version_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface AutomationRunStep {
  automation_run_step_id: string;
  automation_action_id: string;
  position: number;
  status: string;
  result_json: Record<string, unknown> | null;
  has_undo: boolean;
}

export interface AutomationApproval {
  automation_approval_id: string;
  automation_run_id: string;
  automation_run_step_id: string;
  approval_state: string;
  preview_json: Record<string, unknown> | null;
  requested_at: string;
  decided_at: string | null;
  decided_by: string | null;
  approval_reason: string | null;
}

export interface AutomationRun {
  automation_run_id: string;
  automation_workflow_id: string;
  project_id: string;
  trigger_type: string;
  status: string;
  idempotency_key: string | null;
  started_at: string;
  finished_at: string | null;
  summary: string | null;
  steps: AutomationRunStep[];
  approvals: AutomationApproval[];
}

export interface RecipeInstance {
  automation_recipe_instance_id: string;
  recipe_key: string;
  automation_workflow_id: string;
  instance_status: string;
  deferred_capabilities: string[];
  params: Record<string, unknown>;
  created_at: string;
}

export async function getGlobalAutomationSettings(): Promise<GlobalAutomationSettings> {
  return (await requestBackend<GlobalAutomationSettings>("/api/v1/automation/settings")).data;
}

export async function setGlobalAutomationEnabled(globalEnabled: boolean): Promise<GlobalAutomationSettings> {
  return (await requestBackend<GlobalAutomationSettings>("/api/v1/automation/settings", jsonRequest({ global_enabled: globalEnabled }, { method: "PATCH" }))).data;
}

export async function listAutomationRecipes(): Promise<AutomationRecipe[]> {
  return (await requestBackend<AutomationRecipe[]>("/api/v1/automation/recipes")).data;
}

export async function listAutomationWorkflows(projectId: string): Promise<AutomationWorkflow[]> {
  return (await requestBackend<AutomationWorkflow[]>(`/api/v1/projects/${projectId}/automation/workflows`)).data;
}

export async function setWorkflowEnabled(projectId: string, workflowId: string, isEnabled: boolean): Promise<AutomationWorkflow> {
  return (
    await requestBackend<AutomationWorkflow>(
      `/api/v1/projects/${projectId}/automation/workflows/${workflowId}`,
      jsonRequest({ is_enabled: isEnabled }, { method: "PATCH" })
    )
  ).data;
}

export async function listAutomationRuns(projectId: string, limit = 50): Promise<AutomationRun[]> {
  return (await requestBackend<AutomationRun[]>(`/api/v1/projects/${projectId}/automation/runs?limit=${limit}`)).data;
}

export async function getAutomationRun(projectId: string, runId: string): Promise<AutomationRun> {
  return (await requestBackend<AutomationRun>(`/api/v1/projects/${projectId}/automation/runs/${runId}`)).data;
}

export async function runAutomationNow(projectId: string, workflowId: string, idempotencyKey?: string | null): Promise<Record<string, unknown>> {
  return (
    await requestBackend<Record<string, unknown>>(
      `/api/v1/projects/${projectId}/automation/workflows/${workflowId}/run`,
      jsonRequest({ idempotency_key: idempotencyKey ?? crypto.randomUUID() })
    )
  ).data;
}

export async function instantiateRecipe(
  projectId: string,
  recipeKey: string,
  params?: Record<string, unknown> | null,
  isEnabled = true
): Promise<Record<string, unknown>> {
  return (
    await requestBackend<Record<string, unknown>>(
      `/api/v1/projects/${projectId}/automation/recipes/${recipeKey}`,
      jsonRequest({ params: params ?? {}, is_enabled: isEnabled })
    )
  ).data;
}

export async function listRecipeInstances(projectId: string): Promise<RecipeInstance[]> {
  return (await requestBackend<RecipeInstance[]>(`/api/v1/projects/${projectId}/automation/recipe-instances`)).data;
}

export async function listPendingApprovals(projectId: string): Promise<AutomationApproval[]> {
  return (await requestBackend<AutomationApproval[]>(`/api/v1/projects/${projectId}/automation/approvals/pending`)).data;
}

export async function approveAutomationRun(
  projectId: string,
  runId: string,
  approvalId: string,
  decidedBy?: string | null
): Promise<Record<string, unknown>> {
  return (
    await requestBackend<Record<string, unknown>>(
      `/api/v1/projects/${projectId}/automation/runs/${runId}/approvals/${approvalId}/approve`,
      jsonRequest({ decided_by: decidedBy ?? "atlas-ui" })
    )
  ).data;
}

export async function rejectAutomationRun(
  projectId: string,
  runId: string,
  approvalId: string,
  reason?: string | null,
  decidedBy?: string | null
): Promise<Record<string, unknown>> {
  return (
    await requestBackend<Record<string, unknown>>(
      `/api/v1/projects/${projectId}/automation/runs/${runId}/approvals/${approvalId}/reject`,
      jsonRequest({ reason: reason ?? null, decided_by: decidedBy ?? "atlas-ui" })
    )
  ).data;
}

export async function undoAutomationRunStep(projectId: string, stepId: string): Promise<Record<string, unknown>> {
  return (await requestBackend<Record<string, unknown>>(`/api/v1/projects/${projectId}/automation/run-steps/${stepId}/undo`, jsonRequest({}))).data;
}

/** Sum pending approvals across projects for nav badge. */
export async function countPendingApprovals(projectIds: string[]): Promise<number> {
  let total = 0;
  for (const projectId of projectIds) {
    try {
      const pending = await listPendingApprovals(projectId);
      total += pending.length;
    } catch {
      // ignore per-project failures
    }
  }
  return total;
}
